sql/schemachanger/scplan: allow plan to move to NonRevertible earlier

ajwerner opened the pull request:
This is critical for DROP COLUMN. In `DROP COLUMN` we cannot perform the
primary index swap until the dropping column reaches `DELETE_ONLY`, which is
not revertible. The primary index swap itself is revertible. Given this fact,
we need a mechanism to be able to "promote" revertible operations (operations
which don't destroy information or cause irrevocable visible side effects) to
be grouped with or after non-revertible operations. This commit makes that
happen naturally.

Release note: None

ajwerner committed:
sql/schemachanger/scplan: allow plan to move to NonRevertible earlier

This is critical for DROP COLUMN. In `DROP COLUMN` we cannot perform the
primary index swap until the dropping column reaches `DELETE_ONLY`, which is
not revertible. The primary index swap itself is revertible. Given this fact,
we need a mechanism to be able to "promote" revertible operations (operations
which don't destroy information or cause irrevocable visible side effects) to
be grouped with or after non-revertible operations. This commit makes that
happen naturally.

Release note: None

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -695,6 +695,47 @@ func init() {
+					return aID < bID
+				}),
+			}
+		})
```
FWIW I'm totally fine with these, they are well-motivated and easy to understand.

postamar commented on pkg/sql/schemachanger/scplan/internal/scstage/build.go:
```diff
@@ -563,6 +628,26 @@ func (bc buildContext) nodes(current []scpb.Status) []*screl.Node {
+		rec(n)
+	}
+	return m
+}
```
Ah. This makes sense. 

This logic belongs in the bowels of opgen, though, along with what determines the `revertible` boolean field on each op-edge. You'll find that it's quite similar. 

In fact, perhaps this `revertible` boolean field should be replaced by an enum with three states:
- revertible, failure possible;
- revertible, failure not possible;
- not revertible, failure not possible.

ajwerner replied:
I moved this into opgen in the newly added first commit. It was the right thing to do, but took a little bit of elbow grease. 

I've left two booleans as I found it easier to deal with for now, do you feel strongly?

postamar replied:
No, I don't feel strongly about that.

postamar reviewed (approved):
LGTM but please that can-still-fail determining logic to opgen as suggested, it should be quite straightforward to do.

ajwerner committed:
sql/schemachanger/scplan/internal/opgen: track op type information

This change reworks opgen to require the emit functions to specify the type
of op they'll emit. This allows us to statically determine the op type and
then use that to determine whether there are any remaining ops which might
fail.

Release note: None

ajwerner commented:
bors r+
